marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubInputPartitionReader.java

```diff
@@ -0,0 +1,174 @@
+        }
+
+        if (rows == null || rows.length == 0) {
+            logger.info("Finished reading rows");
+            return false;
+        }
+
+        this.currentRow = objectMapper.readTree(rows[rowIndex++]);
+        return true;
+    }
+
+    /**
+     * Convert the most recently read JSON object into an InternalRow.
+     *
+     * @return
+     */
+    @Override
+    public InternalRow get() {
+        Object[] values = Arrays.stream(schema.fields()).map(field -> {
+            String fieldName = field.name();
+            if (currentRow.has(fieldName) && !"null".equals(currentRow.get(fieldName).asText())) {
+                return readValue(field);
+            }
+            return null;
+        }).toArray();
+
+        Row row = RowFactory.create(values);
+        return RowEncoder.apply(this.schema).toRow(row);
+    }
+
+    @Override
+    public void close() {
+        logger.debug("Closing");
+    }
+
+    private ObjectNode buildEndpointConstants(Map<String, String> options, JsonNode initializationResponse, int partitionNumber) {
+        ObjectNode endpointConstants = objectMapper.createObjectNode();
+        endpointConstants.set("initializationResponse", initializationResponse);
+        endpointConstants.put("partitionNumber", partitionNumber);
+        try {
+            endpointConstants.set("sparkSchema", objectMapper.readTree(schema.json()));
+        } catch (Exception ex) {
+            throw new RuntimeException("Unable to write StructType as JSON; cause: " + ex.getMessage(), ex);
+        }
+        return endpointConstants;
+    }
+
+    private OutputCaller.BulkOutputCaller<InputStream> buildOutputCaller(ObjectNode endpointConstants, ObjectNode endpointState) {
+        InputStreamHandle defaultApi = hubClient.getModulesClient().newJSONDocumentManager()
+            .read("/marklogic-data-hub-spark-connector/readRows.api", new InputStreamHandle());
+
+        OutputCaller<InputStream> outputCaller = OutputCaller.on(hubClient.getFinalClient(), defaultApi, new InputStreamHandle());
+
+        return outputCaller.bulkCaller(outputCaller.newCallContext()
+            .withEndpointConstants(new JacksonHandle(endpointConstants))
+            .withEndpointState(new JacksonHandle(endpointState)));
+    }
+
+    private void readNextBatchOfRows() {
+        this.rows = bulkOutputCaller.next();
+        if (rows.length > 0) {
+            numberOfRowsRead += rows.length;
+            if (logger.isDebugEnabled()) {
+                logger.debug("Rows read so far: " + numberOfRowsRead);
+            }
+            rowIndex = 0;
+        }
+    }
+
+    /**
+     * Just doing the bare minimum here for now, will have lots more to support soon.
+     *
+     * @param field
+     * @return
+     */
+    private Object readValue(StructField field) {
+        final String fieldName = field.name();
+        Object value;
+        switch (field.dataType().typeName()) {
+            case "integer":
```

Comment:
This is only supporting what's need for the Customer view in the reference-entity-model project. Will be expanded via a separate story.